### ISSUE TITLE
ood-portal-generator needs a -f when we (developers) update the ood portal template

### ISSUE THREAD

**Issue by johrstrom** [OSC/ondemand#382]
Consider the issue found in [this topic](https://discourse.osc.edu/t/checksum-warning-during-1-6-17-upgrade/520) where _we_ changed a string within the ood portal template, then when checking the check sums, we had to force an update with the ood-portal-generator. 

This is because the checksum changed becuase the ood-portal template in the source code changed.

It seems that the enhancement we made to not overwrite users' hand updates to the `ood-portal.conf` also stops us from updating when _we_ make updates to the source templat.  The checksums will differ not because users/admins updated the resulting file, but because we updated the source file.

**Comment by treydock** [comment]
That shouldn't be the case. I am building a custom version of OnDemand RPM with #370 which changes the template to see if this really happens on a live system. The logic is fairly clear:

```
force || (! File.exist?(apache)) || checksum_matches?(apache)
```

If the ood-portal.conf checksum (saved to disk before dev changes) matches the previously generated checksum from generator, then -f is not needed. This is a way to tell the generator if the checksum last saved from previous execution has changed. Does not take into consideration the template we deploy.

If you look at the `update_ood_portal` function, the generated/templated output is written to a temp file and isn't use to check the checksum.

**Comment by treydock** [comment]
Something isn't right. I built 1.7.7_rc1.1 with #370 and got this during RPM update:

```
WARNING: Checksum of /opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf does not match previous value, not replacing.
Generating new Apache config at: '/opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf.new'
```

However the checksums match:

```
 sha256sum /opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf
25e7cc7af0d888e2fd255342e61580df1c9812fd7be633bf5898282f299d0438  /opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf
```

```
[root@webdev02 ondemand-test]# cat /etc/ood/config/ood_portal.sha256sum 
25e7cc7af0d888e2fd255342e61580df1c9812fd7be633bf5898282f299d0438  /opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf
```

Somehow the checksum at OSC on webdev02 is the checksum of the file without omitting comments.

**Comment by treydock** [comment]
Testing with webdev02 may not work because I don't use `update_ood_portal` to generate our configs. I write them to /etc/ and then have Puppet copy the configs, so the checksums never get updated.

**Comment by treydock** [comment]
Just tested in Vagrant. Had image with 1.7.7_rc1 installed and verified my checksum matched before update then installed 1.7.7_rc1.1 with #370 and the configs got updated.

Before update:

```
[root@ood vagrant]# egrep -v "^\s*#" /opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf | sha256sum
8f4b132db04e928c2070ff108a28e9587d59f1b5a6fcb0340aff857a24415194  -
[root@ood vagrant]# cat /etc/ood/config/ood_portal.sha256sum 
8f4b132db04e928c2070ff108a28e9587d59f1b5a6fcb0340aff857a24415194 /opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf
```

Update text from RPM update:

```
Backing up previous Apache config to: '/opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf.20200113T210541'
Generating new Apache config at: '/opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf'
Generating Apache config checksum file: '/etc/ood/config/ood_portal.sha256sum'
```

New checksum got written after update:

```
[root@ood vagrant]# egrep -v "^\s*#" /opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf | sha256sum
ccb6b87a2077dae62cc7a498dd06102c6589bdc178d6af4a8ff5d250f479e535  -
[root@ood vagrant]# cat /etc/ood/config/ood_portal.sha256sum 
ccb6b87a2077dae62cc7a498dd06102c6589bdc178d6af4a8ff5d250f479e535 /opt/rh/httpd24/root/etc/httpd/conf.d/ood-portal.conf
```

**Comment by johrstrom** [comment]
If I start with a fresh centos:7 container, I can replicate with these steps. I'll have to try again with 1.6.20 to 1.7.x.  I just used 1.6.17 because that's what is in that discourse topic described.

```bash
# from a fresh centos:7 container
rpm -ivh https://yum.osc.edu/ondemand/1.6/ondemand-release-web-1.6-4.noarch.rpm
yum install centos-release-scl
yum install ondemand-1.6.7
# you see the errors here after this update
yum update ondemand-1.6.17

# now add the ci repo
cp /etc/yum.repos.d/ondemand-web.repo /etc/yum.repos.d/ondemand-ci.repo
# edit the new file to use the ci url instead of 1.6

# update again and you'll see the error
yum update ondemand-1.7.7_rc3
```

**Comment by treydock** [comment]
We fixed the behavior in 1.6.20 which is the latest public release.

**Comment by johrstrom** [comment]
Yea [this topic](https://discourse.osc.edu/t/verifying-a-few-details-about-rstudio-app-setup-configuration/664) was also opened with that warning, so I opened this thinking the source template was somehow involved. After looking over the logic a bunch the source template is never in question.  It's always about the result and the previous result. 

Sorry for the confusion!
